colcon_sanitizer_reports/_sanitizer_section_part.py

```diff
@@ -0,0 +1,135 @@
+            <snip irrelevant stack trace>
+
+        SUMMARY: ThreadSanitizer: lock-order-inversion (potential deadlock) <snip>
+    or
+        ==5504==ERROR: LeakSanitizer: detected memory leaks
+
+        Direct leak of 1 byte(s) in 1 object(s) allocated from:
+            <snip relevant stack trace>
+
+        Indirect leak of 1 byte(s) in 1 object(s) allocated from:
+            <snip irrelevant stack trace>
+
+        SUMMARY: AddressSanitizer: 554924 byte(s) leaked in 6385 allocation(s).
+
+    The first example has two section parts. Part one starts with the lock-order-inversion line and
+    includes all the indented lines until the summary line. It includes two relevant and one
+    irrelevant stack traces to report. Part two is the summary line.
+
+    The second example has four section parts. Part one is just the summary line. Parts two starts
+    with the non-indented "Direct leak" line and includes the following stack trace that is relevant
+    to report. Part three starts with the non-indented "Indirect leak" line and includes the
+    following stack trace that is irrelevant to report. The final part is the summary line.
+
+    See _FIND_RELEVANT_STACK_TRACE_BEGIN_REGEXES_BY_ERROR_NAME for stack trace header search
+    patterns that determine which stack traces are relevant. Different error/warning names have
+    different relevant stack traces.
+
+    After initialization, SanitizerSectionPart includes the following data member.
+
+    relevant_stack_traces:
+        Tuple of all stack traces from the section part that are relevant for generating the report.
+    """
+
+    relevant_stack_traces: Tuple[SanitizerSectionPartStackTrace]
+
+    def __init__(self, *, error_name: str, lines: Tuple[str]) -> None:
+        relevant_stack_traces: List[SanitizerSectionPartStackTrace] = []
+        relevant_stack_trace_lines: Optional[List[str]] = None
+        find_relevant_stack_trace_begin_regexes = (
+            _FIND_RELEVANT_STACK_TRACE_BEGIN_REGEXES_BY_ERROR_NAME[error_name]
+        )
+
+        for line in lines:
+            # Check if we're currently gathering lines of a relevant stack trace.
+            if relevant_stack_trace_lines is not None:
+                match = _FIND_STACK_TRACE_LINE_REGEX.match(line)
+                if match is not None:
+                    relevant_stack_trace_lines.append(line)
+                    continue
+
+                relevant_stack_traces.append(
+                    SanitizerSectionPartStackTrace(lines=tuple(relevant_stack_trace_lines))
+                )
+                relevant_stack_trace_lines = None
+                if len(relevant_stack_traces) == len(find_relevant_stack_trace_begin_regexes):
+                    break
+
+            find_relevant_stack_trace_begin_regex = (
+                find_relevant_stack_trace_begin_regexes[len(relevant_stack_traces)]
+            )
+            match = find_relevant_stack_trace_begin_regex.match(line)
```

Comment:
Is it possible for us to miss a beginning of a new stack trace because it matches the `_FIND_STACK_TRACE_LINE_REGEX` and are no non-`_FIND_STACK_TRACE_LINE_REGEX`-matching lines between them?

Reply:
If I understand the scenario you're suggesting correctly, then I don't think we can miss the beginning of a new stack trace. I don't know of any "stack trace begin" pattern that would start with a "#", for instance.

Reply:
Sounds good. I was just trying to think of ways this control flow could violate expectations.

*nuclearsandwich idly wonders if there's a way to validate two regular expressions are mutually exclusive*.